addon/utils/query-cache.js

```diff
@@ -0,0 +1,69 @@
+
+const CACHE_DECAY_TIMEOUT = 5000;
+
+export default class QueryCache {
+  constructor() {
+    this._internalCache = Object.create(null);
+    this.decayId = null;
+  }
+
+  add(type, key, value) {
+    if (!key || !value) {
+      return;
+    }
+
+    if (!this._internalCache[type]) {
+      this._internalCache[type] = Object.create(null);
+    }
+
+    this._internalCache[type][key] = value;
+  }
+
+  remove(type, key) {
+    let typeMap = this._internalCache[type];
+
+    if (typeMap && typeMap[key]) {
+      delete typeMap[key];
+    }
+  }
+
+  clear() {
+    this._internalCache = Object.create(null);
+  }
+
+  get(type, key) {
+    if (!this._internalCache[type]) {
+      return;
+    }
+
+    return this._internalCache[type][key];
+  }
+
+  get types() {
+    return Object.keys(this._internalCache);
+  }
+
+  scheduleDecay(delay = CACHE_DECAY_TIMEOUT, onDecay) {
```

Comment:
> IMO, a more flexible approach would be to handle decay per cache entry rather than wiping out the entire cache.

You're referring to type, not entry, correct?

It seems reasonable to provide a separate eviction strategy for each, but feels odd to be passing that every time in `queryRecord`. This sounds more like something that you'd setup once, up front, then take advantage of repeatedly during the lifetime of the Route. 

For instance, would it not be better to provide a method that allows you to set the decay by type up front, say on `init`?

```js
this.queryCache.setDecay('my-model', () => { ... });
```
This adds more complexity in that you'd need some way of triggering this decay. It feels like we're moving things well into the LRU territory, which we explicitly wanted to avoid here. Caching eviction is very hard to get right. 

> There are 2 hard problems in computer science: cache invalidation, naming things, and off-by-1 errors.